compiler/rustc_mir_transform/src/simplify_static_switch.rs

```diff
@@ -0,0 +1,155 @@
+/// discriminant.
+///
+/// For example, the following CFG:
+/// ```text
+///     x = Foo::A(y); ---       Foo::A ---> ...
+///    /                  \            /
+/// ...                    --> switch x
+///    \                  /            \
+///     x = Foo::B(y); ---       Foo::B ---> ...
+/// ```
+/// would become:
+/// ```text
+///     x = Foo::A(y); --------- Foo::A ---> ...
+///    /
+/// ...
+///    \
+///     x = Foo::B(y); --------- Foo::B ---> ...
+/// ```
+pub struct SimplifyStaticSwitch;
+
+impl<'tcx> MirPass<'tcx> for SimplifyStaticSwitch {
+    fn name(&self) -> &'static str {
+        "SimplifyStaticSwitch"
+    }
+
+    fn is_enabled(&self, sess: &Session) -> bool {
+        sess.mir_opt_level() >= 2
+    }
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        debug!(body = %tcx.def_path_str(body.source.def_id()));
+
+        if simplify_static_switch(tcx, body) {
+            remove_dead_blocks(tcx, body);
+        }
+    }
+}
+
+#[instrument(level = "debug", skip(tcx, body), ret)]
+fn simplify_static_switch<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) -> bool {
+    let mut new_targets = FxHashMap::default();
+    let predecessors: &IndexSlice<_, _> = body.basic_blocks.predecessors();
+    for (block, data) in traversal::preorder(body) {
+        if data.is_cleanup {
+            continue;
+        }
+
+        let predecessors: &[BasicBlock] = &predecessors[block];
+        if predecessors.is_empty() {
+            continue;
+        }
+
+        let TerminatorKind::SwitchInt {
+            discr: Operand::Move(switched) | Operand::Copy(switched),
+            ref targets
+        } = data.terminator().kind else {
+            continue
+        };
+
+        let Some(discr) = data.statements.iter().rev().find_map(|statement| {
+            match statement.kind {
+                StatementKind::Assign(box (place, Rvalue::Discriminant(discr)))
+                    if place == switched =>
+                {
+                    Some(discr)
+                },
+                _ => None,
+            }
+        }) else {
+            continue
+        };
+
+        for &pred in predecessors {
+            let data = &body.basic_blocks[pred];
+            if let TerminatorKind::Goto { .. } = data.terminator().kind {
+                for statement in data.statements.iter().rev() {
+                    match statement.kind {
+                        StatementKind::Assign(box (
+                            place,
+                            Rvalue::Aggregate(box AggregateKind::Adt(def_id, variant, ..), ..),
+                        )) if place == discr => {
+                            debug!(
+                                "{pred:?}: {place:?} = {}::{variant:?}; goto -> {block:?};",
+                                tcx.def_path_str(def_id)
+                            );
+                            let discr_ty = tcx.type_of(def_id).skip_binder();
```

Comment:
You must take the substitutions into account here, otherwise you'll get the wrong type.